src/services/BidService.ts

```diff
@@ -0,0 +1,122 @@
+    await BidModel.createBid({
+      linked_proposal_id,
+      bid_proposal_data,
+      author_address,
+      publish_at,
+      status: BidStatus.Pending,
+    })
+  }
+
+  static async publishBids(context: JobContext) {
+    const pendingBids = await BidModel.getBidsReadyToPublish()
+    if (pendingBids.length === 0) {
+      return
+    }
+
+    if (isNaN(MINIMUM_BIDS_TO_PUBLISH) || MINIMUM_BIDS_TO_PUBLISH === 0) {
+      const errorMsg = `Minimum bids to publish is not set. Please set MINIMUM_BIDS_TO_PUBLISH env variable`
+      ErrorService.report(errorMsg, { category: ErrorCategory.Bid })
+      context.log(errorMsg)
+      return
+    }
+
+    const bidsByTender = new Map<string, typeof pendingBids>()
+    pendingBids.reduce((acc, bid) => {
+      const bids = acc.get(bid.linked_proposal_id) || []
+      bids.push(bid)
+      acc.set(bid.linked_proposal_id, bids)
+      return acc
+    }, bidsByTender)
+
+    const tendersWithBidsToReject = []
+
+    for (const [tenderId, bids] of bidsByTender) {
+      if (bids.length < MINIMUM_BIDS_TO_PUBLISH) {
+        tendersWithBidsToReject.push(tenderId)
+        continue
+      }
+
+      for (const bid of bids) {
+        const { author_address, bid_proposal_data, linked_proposal_id, publish_at } = bid
+        const finish_at = Time(publish_at).add(Number(process.env.DURATION_BID), 'seconds').toDate()
+        try {
+          await ProposalService.createProposal({
+            type: ProposalType.Bid,
+            user: author_address,
+            configuration: {
+              linked_proposal_id,
+              ...bid_proposal_data,
+              choices: DEFAULT_CHOICES,
+            },
+            required_to_pass: 1200000 + bid_proposal_data.funding * 40, // Grants high tier formula
+            finish_at,
+          })
+          await BidModel.removePendingBid(author_address, linked_proposal_id)
+          context.log(`Bid from ${author_address} for tender ${linked_proposal_id} published`)
+        } catch (error) {
+          const msg = `Error publishing bid from ${author_address} for tender ${linked_proposal_id}`
+          ErrorService.report(msg, { error, category: ErrorCategory.Bid })
+          context.log(msg)
+        }
+      }
+    }
+
+    if (tendersWithBidsToReject.length > 0) {
+      try {
+        await BidModel.rejectBidsFromTenders(tendersWithBidsToReject)
+        context.log(`Rejected bids from tenders ${tendersWithBidsToReject.join(', ')}`)
+      } catch (error) {
+        const msg = `Error rejecting bids from tenders ${tendersWithBidsToReject.join(', ')}`
+        ErrorService.report(msg, { error, category: ErrorCategory.Bid })
+        context.log(msg)
+      }
+    }
+  }
+
+  static async hasUserBidOnTender(user: string, tenderId: string) {
+    const bids = await BidModel.getBidsInfoByTender(tenderId)
+    return bids.find((bid) => bid.author_address === user)
+  }
+
+  static async isSubmissionWindowFinished(tenderId: string) {
```

Comment:
I don't get this, we are deciding if a tender is finished based on the amount of rejected bids?

Shouldn't we grab the first bid created for that tender, and add the window duration to the creation date?

🤔

Reply:
The submission window is closed if:

- There is a bid that has already been rejected
- There are bids that have been published

I missed this last condition, I already added it.